ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,433 @@
+Readonly::Scalar my $BIN_DPKG_QUERY => "/usr/bin/dpkg-query";
+Readonly::Scalar my $CMD_APT_UPDATE => [$BIN_APT_GET, qw(-qq update)];
+Readonly::Scalar my $CMD_APT_UPGRADE => [$BIN_APT_GET, qw(-qq dist-upgrade)];
+Readonly::Scalar my $CMD_APT_INSTALL => [$BIN_APT_GET, qw(-qq install)];
+Readonly::Scalar my $CMD_APT_AUTOREMOVE => [$BIN_APT_GET, qw(-qq autoremove)];
+Readonly::Scalar my $CMD_APT_MARK => [$BIN_APT_MARK, qw(-qq)];
+Readonly::Scalar my $CMD_DPKG_QUERY => [$BIN_DPKG_QUERY, qw(-W -f=${db:Status-Abbrev};${Package}\n)];
+
+our $NoActionSupported = 1;
+
+
+# If user packages are not allowed, removes any repositories present
+# in the system that are not listed in $allowed_repos.
+sub cleanup_old_repos
+{
+    my ($self, $repo_dir, $allowed_repos, $allow_user_pkgs) = @_;
+    $self->debug(5, 'Entered cleanup_old_repos()');
+
+    return 1 if $allow_user_pkgs;
+
+    my $dir;
+    if (!opendir($dir, $repo_dir)) {
+        $self->error("Unable to read repositories in $repo_dir");
+        return 0;
+    }
+
+    my $current = Set::Scalar->new(map(m{(.*)\.list$}, readdir($dir)));
+
+    closedir($dir);
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_repos));
+    my $rm = $current-$allowed;
+    foreach my $i (@$rm) {
+        # We use $f here to make Devel::Cover happy
+        my $f = "$repo_dir/$i.list";
+        $self->verbose("Unlinking outdated repository $f");
+        if (!unlink($f)) {
+            $self->error("Unable to remove outdated repository $i: $!");
+            return 0;
+        }
+    }
+    return 1;
+}
+
+
+# Creates the repository dir if needed.
+sub initialize_repos_dir
+{
+    my ($self, $repo_dir) = @_;
+    $self->debug(5, 'Entered initialize_repos_dir()');
+
+    if (! -d $repo_dir) {
+        $self->verbose("$repo_dir didn't exist. Creating it");
+        if (!eval{mkpath($repo_dir)} || $@) {
+            $self->error("Unable to create repository dir $repo_dir: $@");
+            return 0;
+        }
+    }
+    return 1;
+}
+
+
+# Returns the URLs of the repositories listed in $prots, with their
+# host part replaced by each of the reverse @proxies.
+sub generate_reverse_proxy_urls
+{
+    my ($self, $prots, @proxies) = @_;
+    $self->debug(5, 'Entered generate_reverse_proxy_urls()');
+
+    my @l;
+    foreach my $pt (@$prots) {
+        foreach my $px (@proxies) {
+            my $url = $pt->{url};
+            $url =~ s{^(.*?):(/+)[^/]+}{$1:$2$px};
+            push(@l, { url => $url });
+        }
+    }
+    return \@l;
+}
+
+
+# Generate the URLs for the forward proxies, based on whether the
+# repository is accessible over HTTP or HTTPS.
+sub generate_forward_proxy_urls
+{
+    my ($self, $proto, @px) = @_;
+    $self->debug(5, 'Entered generate_forward_proxy_urls()');
+
+    $proto->[0]->{url} =~ m{^(.*?):};
+    return "$1://$px[0]";
+}
+
+
+# Generates the repository files in $repos_dir based on the contents
+# of the $repos subtree. It uses Template::Toolkit $template to render
```

Comment:
it should use `CCM::TextRender`, the `template` moethdo was removed in https://github.com/quattor/ncm-ncd/pull/68
see https://github.com/quattor/configuration-modules-core/pull/808 for an example